oci/layer/generate.go

```diff
@@ -87,21 +90,36 @@ func GenerateLayer(path string, deltas []mtree.InodeDelta, opt *RepackOptions) (
 			switch delta.Type() {
 			case mtree.Modified, mtree.Extra:
 				if packOptions.TranslateOverlayWhiteouts {
-					fi, err := os.Stat(fullPath)
+					woType, isWo, err := isOverlayWhiteout(fullPath, fsEval)
 					if err != nil {
-						return fmt.Errorf("couldn't determine overlay whiteout for %s: %w", fullPath, err)
+						return fmt.Errorf("check if %q is a whiteout: %w", fullPath, err)
 					}
-
-					whiteout, err := isOverlayWhiteout(fi)
-					if err != nil {
-						return err
-					}
-					if whiteout {
-						if err := tg.AddWhiteout(fullPath); err != nil {
-							return fmt.Errorf("generate whiteout from overlayfs: %w", err)
+					if isWo {
+						log.Debugf("generate layer: converting overlayfs whiteout %s %q to OCI whiteout", woType, name)
+
+						var err error
+						switch woType {
+						case overlayWhiteoutPlain:
+							err = tg.AddWhiteout(name)
+						case overlayWhiteoutOpaque:
+							// For opaque whiteout directories we need to
+							// output an entry for the directory itself so that
+							// the ownership and modes set on the directory are
+							// included in the archive.
+							if err := tg.AddFile(name, fullPath); err != nil {
+								log.Warnf("generate layer: could not add directory entry for opaque from overlayfs for file %q: %s", name, err)
+								return fmt.Errorf("generate directory entry for opaque whiteout from overlayfs: %w", err)
+							}
+							err = tg.AddOpaqueWhiteout(name)
+						default:
+							return fmt.Errorf("[internal error] unknown overlayfs whiteout type %q", woType)
+						}
+						if err != nil {
+							log.Warnf("generate layer: could not add whiteout %s from overlayfs for file %q: %s", woType, name, err)
+							return fmt.Errorf("generate whiteout %s from overlayfs: %w", woType, err)
```

Comment:
Is there a reason to both warn and error here and above?

Reply:
We do this in a couple of places, the main issue is that I have seen cases where `CloseWithError` doesn't end up propagating the error properly and so this is meant to be a "final fallback" to tell CLI users that something was up rather than succeeding silently.

I don't know if the `CloseWithError` thing has been fixed. It was a weird race condition that I couldn't ever really reproduce.